## Introduction

Accessible chromatin regions (ACRs) are known to play a critical role in eukaryotic gene regulation but their comprehensive identification in plants remains a challenge [@doi:10.1016/j.bbagrm.2016.06.006;@doi:10.1016/j.tplants.2016.07.013].
Current methods to assay chromatin accessibility are highly environment-specific and relatively expensive compared to DNA sequencing, limiting the number of species or conditions that can be investigated.
Assaying chromatin accessibility in plants comes with additional unique challenges: the cell wall makes plant nuclei hard to isolate and many active transposon families shuffle, create, and destroy regulatory regions over time [@doi:10.1016/j.bbagrm.2016.05.010].
Regions that lack DNA methylation are known to be stable over developmental time and overlap significantly with ACRs in plants with larger genomes [@doi:10.1073/pnas.2010250117], suggesting they may contain a superset of ACRs across cell-types.
Computational models capable of predicting chromatin accessibility and methylation state directly from DNA sequence would enable a wide range of previously-intractable studies on gene regulation across evolutionary time as well as estimation of non-coding variant effects for use in contexts such as breeding.
Plants also provide an excellent system to study the genetic basis of adaptation [@doi:10.1016/j.tig.2011.04.001].
Now that it is feasible to assemble genomes of thousands of species, regulatory regions that control adaptation can be identified, providing valuable insight on how to breed crops resilient to climate change.
Recent advances in machine learning, particularly deep learning, have catalyzed a vast number of applications to biological prediction, including mRNA abundance [@doi:10.1073/pnas.1814551116;@doi:10.1016/j.celrep.2020.107663;@doi:10.1101/2021.04.07.438649], chromatin state [@doi:10.1038/nmeth.3547;@doi:10.1093/nar/gkw226;@doi:10.1371/journal.pcbi.1008050], and transcription factor (TF) binding [@doi:10.1038/s41467-020-18832-8] directly from DNA sequence.
Many of these models have so far only been trained within a single species to predict within the same species, usually utilizing held-out chromosomes as a test set to control for sequence relatedness.

At a high level, plant chromatin has characteristics similar to animal chromatin: chromatin is organized into hierarchical compartments, distal regulatory regions are colocalized to genes through chromatin looping, and various histone modifications signal a wide variety of local chromatin states.
However, the exact mechanisms driving chromatin accessibility are known to be quite different in terms of specific histone modifications [@doi:10.1038/s41477-019-0548-z], pioneer factors [@doi:10.3389/fpls.2021.701406], and chromatin looping mechanisms [@doi:10.1038/s41477-018-0199-5].
Because of these differences, plant-specific chromatin accessibility models are likely to be necessary.

We know that transcription factor binding sites are strongly conserved across evolutionary time [@doi:10.1101/2020.01.07.898056;@doi:10.1371/journal.pcbi.1006484] and highly enriched in ACRs [@doi:10.1038/nrg3682].
Certain deep learning model architectures, such as convolutional neural networks (CNN), have already been shown effective for predicting chromatin accessibility within species by recognizing important motifs [@doi:10.1038/nmeth.3547;@doi:10.1093/nar/gkw226] and their spatial relationships [@doi:10.1038/s41588-021-00782-6].
These CNN-based architectures can accurately predict chromatin accessibility in humans [@doi:10.1038/nmeth.3547;@doi:10.1093/nar/gkw226;@doi:10.1093/bioinformatics/btx679] as well as in plants [@doi:10.1093/nar/gkab383;@doi:10.1016/j.gpb.2020.06.021].
However, the vast majority of previous work has focused on improving performance within species and across cell-types, with little focus on across-species prediction [@doi:10.1371/journal.pcbi.1008050] or prediction within unobserved species.
Previous work [@doi:10.1101/2020.07.28.224485;@doi:10.1371/journal.pcbi.1006484] has observed that CNNs require much larger training data sets than earlier model architectures to achieve equivalent or better performance.
By incorporating multiple species into the training data we not only increase the number of observations but also the total evolutionary time between observations, which reduces confounding neutral variation within conserved sequences.
For the purposes of predicting regulatory regions in unobserved plant species, training a model across species will be critical to learn important motifs and syntax that are conserved across longer evolutionary time periods.
Therefore, we predicted that previously-published deep learning architectures could work well across species and make accurate chromatin accessibility and methylation predictions in related unobserved species.

Here, we train DanQ [@doi:10.1093/nar/gkw226] to predict chromatin accessibility using leaf ATAC-seq data from 12 angiosperm species [@doi:10.1038/s41477-019-0548-z], comparing the performance of within-species models to across-species models.
We also train DanQ to predict unmethylated regions using methylation data from 10 angiosperm species, including 5 previously-published grasses [@doi:10.1073/pnas.2010250117].
Using a maize single-cell ATAC (scATAC) accessibility atlas [@doi:10.1016/j.cell.2021.04.014], we see that the accessibility model has similar performance across cell-types but is highly variable across regions with different levels of cell-type specificity.
Using various interpretation methods designed for CNNs, we compare and contrast which motifs were important across angiosperms for predicting chromatin accessibility in leaves or methylation state.
Our publicly available pan-angiosperm chromatin state models are an important stepping stone towards a better understanding of gene regulation and adaptation.
