## Results

### Recurrent convolutional neural networks accurately model chromatin state across species

To train a successful chromatin state classifier, we needed to choose a window size that balanced genomic context with resolution.
We tested a few different model configurations and decided upon 600 base pair windows because higher window sizes showed diminishing returns on performance on our validation set while decreasing our effective resolution (Figure {@fig:model-configs}).
We preprocessed the ATAC-Seq and unmethylated peaks by taking the midpoint and symmetrically extending to half the window size in both directions to obtain our positive observations.
Negatives were sampled from the rest of the genome.
After preprocessing we had 26,280 training regions per species (315,360 total) for the across-species accessibility models and 35,652 training regions per species (356,520 total) for the methylation models, split evenly between classes.

![Performance of the across-species chromatin state classifiers. The top middle and top right show the mean and standard error (due to variability in the stochastic model parameter initialization processes) of the auPR for the accessibility and methylation models, respectively, per species for both the within- and across-species training configurations. The bottom left is the precision-recall curve across all hold-out species for the across-species models, split by distance class and chromatin feature. The bottom middle and bottom right are the precision-recall curves for the across-species accessibility and methylation models, respectively, split by species.](images/fig1.png){#fig:performance}

As a baseline for comparison to previous, within-species, chromatin state CNN models as well as our across-species models, we trained within-species DanQ model configurations for each of the angiosperm species in our data.
We also trained across-species model configurations each using a different species as a test set.
Generally, we observed that a given across-species model has a comparable, if not superior, area under the precision-recall curve (auPR) to the within-species model (Figure {@fig:performance}, top middle and top right).
While the across-species accessibility model auPR and areas under the receiver operating characteristic curve vary substantially (Figure {@fig:performance}, bottom middle and bottom right, and Figure {@fig:roc}), they are also within the range of those observed in the original DanQ and DeepSEA human models and superior to the bag-of-kmers model within _Zea mays_ (Figure {@fig:bok-comparison}).
We also see that both within-species and across-species performance decreases as genome size increases (Figure {@fig:genome-size-auPR}).
When comparing the accessibility and methylation models, we see the same trends in performance for each species.
To assess whether recurrent CNNs were a better architecture choice for across-species accessibility models over standard CNNs, we trained across-species accessibility configurations of the Basset, CharPlant, and DeeperDeepSEA architectures.
We observed that DanQ was a superior architecture for across-species accessibility modeling for almost all hold-out species (Figure {@fig:model-cmp}).

To see if the models were more accurate in predicting accessible or unmethylated regions near or within genes, where these regions are known to be enriched, we looked at the precision-recall curves across different distance classes (genic, proximal, or distal).
Observations were labeled as genic if more than half of the range overlapped with a gene annotation, as proximal if not genic and more than half of the range was within the proximal cutoff (2kb), and as distal if neither genic nor proximal.
We see that the across-species models for both chromatin features perform the worst on distal regions, but show contrasting results on the genic and proximal regions (Figure {@fig:performance}, bottom left).
This could be driven by the imbalanced distribution of regions between the distance classes, with accessible regions biased towards the proximal class and unmethylated regions towards the genic class (Figure {@fig:distance_class_species_count}).
In particular, _Hordeum vulgare_ has proportionally many more distal accessible and unmethylated regions, which could explain the lower overall performance.
The across-species accessibility models are very precise when calling inaccessible chromatin, with most of the errors being false-positives, particularly in distal regions (Figure {@fig:fdr-vs-for}).
We see a much different result in the methylation model, which shows only a slight bias towards false positives.

To control for potential _trans_-driven transposon silencing, we tested a two-step model that takes the predictions of the a2z model and then masks them with zeros if they overlap annotated transposons in _Z. mays_.
We see that these two-step repeat-masked models do much better (Î”auPR 0.15 for accessibility and 0.07 for methylation) than the naive models (Figure {@fig:maize-masked}), suggesting a relatively straightforward way to reduce false positives in larger plant genomes with more transposon-derived sequence.

Finally, we wanted to assess how far out in evolutionary time the angiosperm model could work.
We ran the model against ATAC-Seq data from _Saccharomyces cerevisiae_ and a _Homo sapiens_ GM12878 cell line [@doi:10.1101/gr.192294.115].
We see the plant-trained model has some ability (Figure {@fig:euk-cmp}) to predict chromatin accessibility in yeast (auPR 0.21), if not human cell-lines (auPR 0.02).

### Leaf-trained models struggle to predict cell type-specific accessible chromatin regions

![Across-cell type performance of the _Zea mays_ accessibility model. The left plot shows the area under the threshold-recall curve for each set of peaks grouped by the number of cell types they are accessible in. The right plot shows the precision-recall curves for peaks accessible in the guard cell (best) and trichoblast (worst) cell types, as well as peaks open in any cell type (union).](images/fig2.png){#fig:cross-tissue}

Knowing the a2z models are capable of working across species, we then asked how well the leaf-trained accessibility models could work across cell types.
We used scATAC-Seq data from six maize organs [@doi:10.1016/j.cell.2021.04.014] as a multi-cell type test set for our single-tissue model.
Using a model trained on every species with ATAC-seq data except _Z. mays_, we predicted the accessibility of each scATAC peak as well as negatives sampled from the rest of the genome.
Looking at the area under the threshold-recall curve we see that the model does better on peaks that are accessible across many cell types, with a sharp decrease in peaks only accessible in five or fewer cell types, which are likely to be a mix of false positives and highly cell type-specific peaks (Figure {@fig:cross-tissue}, left).
The model does best on peaks that are generally open across many cell types, which comprise a largest portion of the training data (Figure {@fig:peak_celltype_specificity}).
This is clearly shown when looking at the overall precision-recall curves in the best (guard cell) and worst (trichoblast) cell types, as well as a union of all cell types.
There is not a substantial difference between the three (Figure {@fig:cross-tissue}, right).

### Interpretation methods reveal important conserved and species-specific motifs

Although chromatin state models that work across angiosperms are a useful tool, we may be able to gain new insights into chromatin biology by dissecting what motifs and higher-order motif patterns the model is learning to use to separate accessible from inaccessible chromatin or unmethylated from methylated regions.
We started with the attribution tool TF-MoDISco to identify important motifs in the _Z. mays_ and _Arabidopsis thaliana_ test sets using their respective held-out models.
While TF-MoDISco qualitatively identified many important motifs (Figure {@fig:tf-modisco}), most of them ranked similarly by attribution score and therefore could not be quantitatively compared in terms of effect size or importance relative to each other.

![Multidimensional scaling of the high-effect medoid kmer distance matrix across all species and chromatin feature model combinations. Each point is a high-effect kmer in a given species and chromatin feature combination.](images/fig4.png){#fig:mds}

To obtain better estimates of sequence effect size, we developed a method that masks sliding windows across a set of sequences and evaluates the change in the model prediction, which we refer to as the kmer occlusion method.
Using a kmer size of 10bp, representing a common estimate of core binding site length, we ran a kmer occlusion to get effect sizes for each kmer in the test set, binned kmers into "high-effect" and "null-effect", and then scanned them for matches to JASPAR 2020 CORE _plantae_ [@doi:10.1093/nar/gkz1001] binding motifs.
For our accessibility models, we see that about 20-40% of high-effect kmers match with JASPAR motifs while our methylation models generally seem to have poor matching between JASPAR motifs and high-effect kmers (Figure {@fig:jaspar-match}).
To look at how similar the high-effect kmers were between chromatin features and species, we used k-medoids to get a subset of representative kmers and then visualized the distances between them using multidimensional scaling.
Surprisingly, the high-effect kmers across species and chromatin features cluster together, with slight separation between methylation and accessibility (Figure {@fig:mds}, left).
However, there is no separation between species (Figure {@fig:mds-species}) nor monocots and dicots (Figure {@fig:mds}, middle/right) for either chromatin feature.

![Positional Global Importance Analysis plots for _A. thaliana_ (left) and _Z. mays_ (right) accessibility (top) and methylation (bottom). The solid and dotted lines represent the importance scores for the positive and negative strand, respectively. Only the top three JASPAR motifs ranked by the maximum global importance across the sequence were plotted.](images/fig3.png){#fig:interpretation-kmers}

To understand which known biological motifs were being recognized as important to the model, we used a recently-developed model interpretation method known as Global Importance Analysis (GIA) [@doi:10.1371/journal.pcbi.1008925].
First, we ranked JASPAR motifs by their max global importance across all positions for each model (Table {@tbl:gia-top10-jaspar}) and see both species-specific and common TFs across the models.
One of the most remarkable observations is that the top 10 motifs in the _A. thaliana_ model are all from the TCP family.
The _Z. mays_ accessibility model also ranked TCP motifs in the top 10 but behind Dof-type motifs.
The _A. thaliana_ and _Z. mays_ methylation models rank the same two motifs at the top and share mostly the same families between the rest.
Next, we looked at the positional effects of the top three TFs across _A. thaliana_ accessibility (Figure {@fig:interpretation-kmers}, top left) and methylation (bottom left) as well as _Z. mays_ accessibility (top right) and methylation (bottom right).
The most striking feature is the sawtooth pattern seen across both species and chromatin feature models, however the cause of this pattern is unclear.
The _A. thaliana_ accessibility model shows a clear bias towards the center of the accessible regions for the top three TFs while the other models are not as consistent.
