## Supplementary Information

![Mean auPR of different across-species accessibility model configurations on the test set.
From left to right: 1000 bp windows, 300 bp windows, 600 bp windows, training and testing within only grasses, using a training set balanced on both accessibility and distance class, exponential activation on the convolutional layer, and testing on Arabidopsis and Maize while training on the rest.](images/supplemental/model_configs.png){#fig:model-configs tag="S1"}

![Receiver operating characteristic curves of the across-species models per hold-out species.
The gray dashed line is the baseline expectation for a random classifier.
The area under the curve is given inside the parentheses for each species in the figure legends.](images/supplemental/roc.png){#fig:roc tag="S2"}

![Precision-recall curve comparison of the across-species a2z model with the bag-of-_k_-mers model in _Z. mays_.
The auPR is shown within parentheses in the figure legends.](images/supplemental/bok-vs-a2z.png){#fig:bok-comparison tag="S3"}

![auPR of both model configurations with respect to genome size.
The dashed line is an exponential fit to the across-species model auPR values.](images/supplemental/genome-size-auPR.png){#fig:genome-size-auPR tag="S4"}

![Performance comparison of different model architectures in the across-species accessibility model training configuration.
Each colored line represents a different test species.
In the left subfigure each point is the average performance of three models of a given architecture trained on all other species and tested within the unobserved species denoted by the point color.
The right subfigure shows the difference in the average performance of each architecture from DanQ for each unobserved test species.
The models are arranged on the x-axis by ascending average auPR across all test species.](images/supplemental/model_cmp.png){#fig:model-cmp tag="S5"}

![Counts of accessible regions in the across-species test sets by distance class and species.](images/supplemental/distance_class_species_count.png){#fig:distance_class_species_count tag="S6"}

![Comparison of false discovery rate (FDR) versus false omission rate (FOR) between models.](images/supplemental/fdr_vs_for.png){#fig:fdr-vs-for tag="S7"}

![Precision-recall curve comparison between a2z model and the repeat-masked a2z model in _Z. mays_.](images/supplemental/maize-masked.png){#fig:maize-masked tag="S8"}

![Precision-recall and receiver operating characteristic curves of an angiosperm-trained a2z model on yeast and a human cell line.](images/supplemental/euk-cmp.png){#fig:euk-cmp tag="S9"}

![Cell type specificity of maize scATAC-Seq peaks.](images/supplemental/peak_celltype_specificity.png){#fig:peak_celltype_specificity tag="S10"}

![Top 3 TF-MoDISco patterns for four a2z models.
_A. thaliana_ is the left column, _Z. mays_ is the right column.
Accessibility is the top row and methylation is the bottom row.
Within each species and chromatin feature combination the patterns are ranked from top to bottom by the number of supporting seqlets for that pattern.](images/supplemental/tf-modisco.png){#fig:tf-modisco tag="S11"}

![Proportion of high-effect _k_-mers that significantly (q-value \<\= 0.05) matched JASPAR CORE _plantae_ motifs with FIMO, grouped by species and chromatin feature.](images/supplemental/jaspar-match.png){#fig:jaspar-match tag="S12"}

![Multidimensional scaling of the top 10 high-effect medoid _k_-mers for each species and chromatin feature model combination, colored by species.](images/supplemental/mds-species.png){#fig:mds-species tag="S13"}

| Accession  | Species              | Tissue | Library  | Coverage (X) |
|:-----------|:---------------------|:------:|:--------:|:------------:|
| SRX096372  | Arabidopsis thaliana |  Leaf  | 101bp SE |      81      |
| SRX1436235 | Eutrema salsugineum  |  Leaf  | 101bp SE |      49      |
| SRX1656936 | Vitis vinifera       |  Leaf  | 101bp SE |      28      |
| SRX1656930 | Populus trichocarpa  |  Leaf  | 101bp SE |      31      |
| SRX200136  | Glycine max          |  Leaf  | 101bp SE |      34      |
| SRX200137  | Glycine max          |  Leaf  | 101bp SE |      27      |

Table: Data used to call dicot unmethylated regions. {#tbl:dicot-umr-data tag="S14"}
